spread/openstack.go

```diff
@@ -0,0 +1,565 @@
+			time.Sleep(2 * time.Second)
+			continue
+		}
+		if server.Status != nova.StatusBuild {
+			if server.Status != nova.StatusActive {
+				return fmt.Errorf("server status is not active: %s", server.Status)
+			}
+			break
+		}
+		time.Sleep(5 * time.Second)
+	}
+	debugf("server %s is running", s.d.Name)
+	return nil
+}
+
+func (p *openstackProvider) waitServerCompleteSetup(s *openstackServer, timeoutSeconds int) error {
+	server, err := p.computeClient.GetServer(s.d.Id)
+	if err != nil {
+		return fmt.Errorf("error retrieving server information: %v", err)
+	}
+	// The adreesses for a network is map of networks and list of ip adresses
+	// We are configuring just 1 network address for the network
+	s.address = server.Addresses[s.d.Network][0].Address
+
+	config := &ssh.ClientConfig{
+		User:            "root",
+		Auth:            []ssh.AuthMethod{ssh.Password(p.options.Password)},
+		Timeout:         10 * time.Second,
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	addr := s.address
+	if !strings.Contains(addr, ":") {
+		addr += ":22"
+	}
+
+	// Iterate until the ssh connection to the host can be stablished
+	// In openstack the client cannot access to the serial console of the instance
+	start := time.Now()
+	for {
+		if time.Since(start) > time.Duration(timeoutSeconds)*time.Second {
+			return &FatalError{fmt.Errorf("failed to ssh to the allocated instance")}
+		}
+
+		_, err = ssh.Dial("tcp", addr, config)
+		if err == nil {
+			break
+		}
+
+		time.Sleep(2 * time.Second)
+	}
+	debugf("connection to server %s is stablished", s.d.Name)
+	return nil
+}
+
+func (p *openstackProvider) createMachine(ctx context.Context, system *System) (*openstackServer, error) {
+	debugf("Creating new openstack server for %s...", system.Name)
+
+	name := openstackName()
+	flavorName := openstackDefaultFlavor
+	if system.Plan != "" {
+		flavorName = system.Plan
+	}
+	flavor, err := p.findFlavor(flavorName)
+	if err != nil {
+		return nil, err
+	}
+
+	network, err := p.findNetwork()
+	if err != nil {
+		return nil, err
+	}
+
+	image, err := p.findImage(system.Image)
+	if err != nil {
+		return nil, err
+	}
+
+	availabilityZone, err := p.findAvailabilityZone()
+	if err != nil {
+		return nil, err
+	}
+
+	// cloud init script
+	cloudconfig := fmt.Sprintf(openstackCloudInitScript, p.options.Password)
+
+	// tags to the created instance
+	tags := map[string]string{
+		"spread":   "true",
+		"owner":    strings.ToLower(username()),
+		"reuse":    strconv.FormatBool(p.options.Reuse),
+		"password": p.options.Password,
+	}
+
+	opts := nova.RunServerOpts{
+		Name:             name,
+		FlavorId:         flavor.Id,
+		ImageId:          image.Id,
+		AvailabilityZone: availabilityZone.Name,
+		Networks:         []nova.ServerNetworks{{NetworkId: network.Id}},
+		Metadata:         tags,
+		UserData:         []byte(cloudconfig),
+	}
+	server, err := p.computeClient.RunServer(opts)
+	if err != nil {
+		return nil, &FatalError{fmt.Errorf("Could not create instance", err)}
```

Comment:
Should the format string include a format specifier here?

Reply:
not needed, thanks

Reply:
the FatalError means it is not needed to retry and it is built from an error, I already updated the error in the backend to make sure we retry only when it makes sense

Reply:
Oh okay, I see know thanks for explaining.